### ISSUE TITLE
Static typing with functools.partial

### ISSUE THREAD

**Issue by benjaminsaljooghi** [microsoft/pyright#220]
**Is your feature request related to a problem? Please describe.**
Calls to functools.partial returns an object of the `partial` class which cannot be used for static typing.

**Describe the solution you'd like**
Pyright analyzes the args and kwargs of calls to functools.partial to perform static typing with the resulting partial object.

```Python
from functools import partial

def add(a: int, b: int) -> int:
    return a + b

add3 = partial(add, 3)

add3("hello") # Pyright should raise an error
```

**Comment by erictraut** [comment]
Thanks for the suggestion.

Pyright is performing correct type checking based on the definition of the `partial` class in the typeshed stub:
```
class partial(Generic[_T]):
    func = ...  # type: Callable[..., _T]
    args = ...  # type: Tuple[Any, ...]
    keywords = ...  # type: Dict[str, Any]
    def __init__(self, func: Callable[..., _T], *args: Any, **kwargs: Any) -> None: ...
    def __call__(self, *args: Any, **kwargs: Any) -> _T: ...
```

Unfortunately the python type language isn't expressive enough to specify the full behavior of partial. The only way to provide more detailed type checking here would be to hard-code knowledge about the internals of the `partial` class within pyright. In general, that's not something I want to do. There have been a few exceptions for very popular classes like `namedtuple`. I don't think `partial` meets the bar though.

If you want static typing in this case, it's trivial to create your own custom class that does what `partial` does but doesn't rely on untyped args and kwargs.

**Comment by codethief** [comment]
@erictraut 

>  The only way to provide more detailed type checking here would be to hard-code knowledge about the internals of the partial class within pyright.

It seems Mypy recently went that way and added a plugin for `functools.partial`, see https://github.com/python/mypy/issues/1484 and https://github.com/python/mypy/pull/16939 . In light of this, would it make sense to revisit the issue?

**Comment by erictraut** [comment]
@codethief, pyright does have extensive special-casing for `functools.partial`. I added this several years ago (after this issue was created) based on feedback from users that this was an important case to handle. In fact, pyright's handling of `functools.partial` is more complete than mypy's at this point. If you're experiencing problems or have questions, feel free to start a new discussion thread.

**Comment by bschnurr** [comment]
yup. see an error now. 

![Image](https://github.com/user-attachments/assets/fe0d7d4e-cb95-4fc3-a3b1-c88da466b8ad)

**Comment by codethief** [comment]
Interesting, I was under the impression it only tracked/type-checked the function's return value. I'll have another look! Thanks so much!
